### PR TITLE
Added CONV2D and DEPTHWISE_CONV2D optimizations for the Vision P6

### PR DESCRIPTION
**Pull request by jackagoogle**
This PR contains a subset of changes from https://github.com/tensorflow/tflite-micro/pull/520

BUG=http://b/189255344

### CODE REVIEW

File: tensorflow/lite/micro/kernels/xtensa/xtensa_conv.h [hunk 1]
```diff
@@ -50,12 +60,10 @@ TfLiteStatus ConvEvalHifi16(TfLiteContext* context, TfLiteNode* node,
                             const TfLiteEvalTensor* filter,
                             const TfLiteEvalTensor* bias,
                             TfLiteEvalTensor* output);
+#else
+  TfLiteStatus ConvReferenceEvalInt16(TfLiteContext* context, TfLiteNode* node);
 #endif  // defined (HIFI4_INTERNAL)
 
-TfLiteStatus ConvReferenceEvalInt8(TfLiteContext* context, TfLiteNode* node);
```
Comment (tingyan19) [inline tensorflow/lite/micro/kernels/xtensa/xtensa_conv.h:55]: Does it help if we keep ConvReferenceEvalInt8() and ConvReferenceEvalInt16() outside the #if define's? It does not affect code size if they're not used, and it reduces burden of future maintenance if we do not have to worry about which #if defined() condition/branch they belong to.

Comment (jackagoogle) [reply tensorflow/lite/micro/kernels/xtensa/xtensa_conv.h:55]: Done.

**Review by tingyan19 — DISMISSED** [review]

File: tensorflow/lite/micro/kernels/xtensa/xtensa_conv.h [hunk 2]
```diff
@@ -52,8 +60,7 @@ TfLiteStatus ConvEvalHifi16(TfLiteContext* context, TfLiteNode* node,
                             TfLiteEvalTensor* output);
 #endif  // defined (HIFI4_INTERNAL)
 
-TfLiteStatus ConvReferenceEvalInt8(TfLiteContext* context, TfLiteNode* node);
-
+TFLiteStatus ConvReferenceEvalInt8(TfLiteContext* context, TfLiteNode* node);
```
Comment (tingyan19) [inline tensorflow/lite/micro/kernels/xtensa/xtensa_conv.h:63]: TfLiteStatus instead of TFLiteStatus?

Comment (jackagoogle) [reply tensorflow/lite/micro/kernels/xtensa/xtensa_conv.h]: Done.

**Review by tingyan19 — APPROVED** [review]

File: tensorflow/lite/micro/kernels/xtensa/depthwise_conv_vision.cc [hunk 3]
```diff
@@ -0,0 +1,135 @@
+#if defined(VISIONP6)
```
Comment (tingyan19) [inline tensorflow/lite/micro/kernels/xtensa/depthwise_conv_vision.cc:1]: Missing copyright info

Please run clang-format and cpplint.py on modified/added files before submitting https://github.com/tensorflow/tflite-micro/blob/main/CONTRIBUTING.md#before-submitting-your-pr

File: tensorflow/lite/micro/kernels/xtensa/conv_vision.cc [hunk 4]
```diff
@@ -0,0 +1,140 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (tingyan19) [inline tensorflow/lite/micro/kernels/xtensa/conv_vision.cc:2]: Please fix formatting issues with the copyright info https://github.com/tensorflow/tflite-micro/runs/4655335928?check_suite_focus=true

File: tensorflow/lite/micro/kernels/xtensa/depthwise_conv_vision.cc (hunk 3 above)
Comment (advaitjain) [reply tensorflow/lite/micro/kernels/xtensa/depthwise_conv_vision.cc:1]: I have pushed a commit with these fixes.

File: tensorflow/lite/micro/kernels/xtensa/conv_vision.cc (hunk 4 above)
Comment (advaitjain) [reply tensorflow/lite/micro/kernels/xtensa/conv_vision.cc:2]: I have pushed a commit with these fixes.

**Review by advaitjain — APPROVED** [review]
approving since I only made formatting changes @tingyan19's last approval.

**Comment by advaitjain** [comment]
Note that the changes from this PR cause the kernel tests to fail.

```
make -f tensorflow/lite/micro/tools/make/Makefile \
    TARGET=xtensa \
    TARGET_ARCH=visionp6 \
    OPTIMIZED_KERNEL_DIR=xtensa \
    XTENSA_CORE=P6_200528 \
    test_kernel_conv_test
```

gives errors like:
```
Testing Int8Filter8x3x3x3PerChannelScaleRelu6ShouldMatchGolden
expected_output_data[i] (-1.2656248*2^6) near output_data[i] (-1.4687496*2^6) failed at tensorflow/lite/micro/kernels/conv_test_common.cc:55
expected_output_data[i] (-1.7031246*2^6) near output_data[i] (-1.8124997*2^6) failed at tensorflow/lite/micro/kernels/conv_test_common.cc:55
expected_output_data[i] (-1.8124997*2^4) near output_data[i] (-1.7499998*2^2) failed at tensorflow/lite/micro/kernels/conv_test_common.cc:55
expected_output_data[i] (1.0*2^2) near output_data[i] (1.4374997*2^4) failed at tensorflow/lite/micro/kernels/conv_test_common.cc:55
```

Vision P6 tests are not part of the per-PR CI: https://github.com/tensorflow/tflite-micro/pull/405
Internal bug: http://b/212719048

**Comment by psharath1** [comment]
unable to reproduce this error.. All tests passing when run 
tensorflow/lite/micro/testing/test_xtensa_binary.sh tensorflow/lite/micro/tools/make/gen/xtensa_visionp6_default/bin/kernel_conv_test '~~~ALL TESTS PASSED~~~' xtensa
Testing FilterDimsNotMatchingAffineQuantization
tensorflow/lite/micro/kernels/conv_common.cc:170 affine_quantization->scale->size == 1 || affine_quantization->scale->size == filter->dims->data[kConvQu was not true.
tensorflow/lite/micro/kernels/conv_common.cc:170 affine_quantization->scale->size == 1 || affine_quantization->scale->size == filter->dims->data[kConvQu was not true.
Testing Int8Input32x1Filter32x32ShouldMatchGolden
Testing Int8Filter8x3x3x3PerChannelScaleRelu6ShouldMatchGolden
3/3 tests passed
~~~ALL TESTS PASSED~~~

Can you please let me know if I need to follow any other steps?

**Comment by tingyan19** [comment]
Can you start from a fresh clone following the following steps? Please let us know if you could reproduce the test errors.

mkdir test

cd test

git clone https://github.com/tensorflow/tflite-micro.git

cd tflite-micro/

git checkout df88e0641143534df1b4eed6d314145263fa52b7

make -f tensorflow/lite/micro/tools/make/Makefile     TARGET=xtensa     TARGET_ARCH=visionp6     OPTIMIZED_KERNEL_DIR=xtensa     XTENSA_CORE=P6_200528     test_kernel_conv_test

**Comment by psharath1** [comment]
> Can you start from a fresh clone following the following steps? Please let us know if you could reproduce the test errors.
> 
> mkdir test
> 
> cd test
> 
> git clone https://github.com/tensorflow/tflite-micro.git
> 
> cd tflite-micro/
> 
> git checkout [df88e06](https://github.com/tensorflow/tflite-micro/commit/df88e0641143534df1b4eed6d314145263fa52b7)
> 
> make -f tensorflow/lite/micro/tools/make/Makefile TARGET=xtensa TARGET_ARCH=visionp6 OPTIMIZED_KERNEL_DIR=xtensa XTENSA_CORE=P6_200528 test_kernel_conv_test

Was able to reproduce the issue with this merge. However, some checks which were added have been removed from this merge leading to the failure. please note that we have supported only the person detect use case in this and hence have restricted the stride to 1 ( and same stride along width and height). The test case failing has a stride of 2. In our original merge, this defaults to the reference (unoptimized) function call and hence passes. 
in this branch, the checks for stride have been removed and hence the optimized function is called with wrong parameters resulting in failure.

**Comment by psharath1** [comment]
Also note that convolution is optimized for 1x1 filter as is the case for person detect example

**Comment by jackagoogle** [comment]
I can add the checks back in, but it gives a false sense of security that an op you have may be optimized without thoroughly digging through. Can you make the tests pass in the example test cases to at least make the kernels slightly more robust?

**Comment by psharath1** [comment]
We are working on providing full support for conv and depthwise convolution.

**Pull request merged**